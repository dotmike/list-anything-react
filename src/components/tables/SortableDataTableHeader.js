import React, { Component } from "react";
import PropTypes from "prop-types";
import { Table } from "semantic-ui-react";

class SortableDataTableHeader extends Component {
  state = {
    column: null,
    direction: null
  };

  componentWillReceiveProps(nextProps) {
    if (this.props.data !== nextProps.data)
      this.setState({
        column: null,
        direction: null
      });
  }

  generateCell = (id, content) => (
    <Table.HeaderCell
      key={id}
      sorted={this.state.column === id ? this.state.direction : null}
      onClick={this.handleSort(id)}
    >
      {content}
    </Table.HeaderCell>
  );

  handleSort = clickedColumn => () => {
    const { column, direction } = this.state;
    const { sort } = this.props;

    if (column !== clickedColumn) {
      sort(column, clickedColumn);
      this.setState({
        column: clickedColumn,
        direction: "ascending"
      });
    } else {
      sort(column, clickedColumn);
      this.setState({
        direction: direction === "ascending" ? "descending" : "ascending"
      });
    }
  };

  render() {
    return (
      <Table.Header>
        <Table.Row>{this.props.renderHeader(this.generateCell)}</Table.Row>
      </Table.Header>
    );
  }
}

SortableDataTableHeader.propTypes = {
  // ownProps
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  sort: PropTypes.func.isRequired,
  renderHeader: PropTypes.func.isRequired
};

export default SortableDataTableHeader;
