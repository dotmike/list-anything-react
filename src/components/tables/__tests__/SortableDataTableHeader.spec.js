import SortableDataTableHeader from "../SortableDataTableHeader";

const testData = {
  data: [{ id: "1", content: "Test" }, { id: "2", content: "My Test" }],
  newDataValue: { id: "3", content: "Test 2" }
};

const setup = (propOverrides = {}) => {
  const props = {
    data: testData.data,
    sort: jest.fn(),
    renderHeader: generateCell =>
      testData.data.map(({ id, content }) => generateCell(id, content)),
    ...propOverrides
  };

  return {
    props,
    wrapperShallow: wrapperShallow(SortableDataTableHeader, props)
  };
};

describe("SortableDataTableHeader", () => {
  it("renders correctly", () => {
    const { wrapperShallow: wrapper } = setup();
    expect(wrapper()).toMatchSnapshot();
  });

  it("calls `sort` when clicking on a `TableHeaderCell`", () => {
    const { wrapperShallow: wrapper, props } = setup();

    wrapper()
      .find("TableHeaderCell")
      .first()
      .simulate("click");

    expect(props.sort).toHaveBeenCalled();
  });

  it("resets the `TableHeaderCell`'s prop `sorted` when receiving new prop `data`", () => {
    const { wrapperShallow: wrapper } = setup();

    wrapper()
      .find("TableHeaderCell")
      .first()
      .simulate("click");

    expect(
      wrapper()
        .find("TableHeaderCell")
        .first()
        .props().sorted
    ).not.toBe(null);

    wrapper().setProps({ data: [...testData.data, testData.newDataValue] });

    expect(
      wrapper()
        .find("TableHeaderCell")
        .first()
        .props().sorted
    ).toBe(null);
  });
});
