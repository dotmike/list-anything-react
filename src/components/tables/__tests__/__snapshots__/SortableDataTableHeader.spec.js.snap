// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortableDataTableHeader renders correctly 1`] = `
<TableHeader
  as="thead"
>
  <TableRow
    as="tr"
    cellAs="td"
  >
    <TableHeaderCell
      as="th"
      key="1"
      onClick={[Function]}
      sorted={null}
    >
      Test
    </TableHeaderCell>
    <TableHeaderCell
      as="th"
      key="2"
      onClick={[Function]}
      sorted={null}
    >
      My Test
    </TableHeaderCell>
  </TableRow>
</TableHeader>
`;
